libfwupdplugin/fu-fmap-firmware.c

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (C) 2020 Benson Leung <bleung@chromium.org>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-fmap-firmware.h"
+
+struct _FuFmapFirmware {
+	FuFirmware		 parent_instance;
+};
+
+G_DEFINE_TYPE (FuFmapFirmware, fu_fmap_firmware, FU_TYPE_FIRMWARE)
+
+/* returns size of fmap data structure if successful, <0 to indicate error */
+static gint
+fmap_size (FuFmap *fmap)
+{
+	if (fmap == NULL)
+		return -1;
+
+	return sizeof (*fmap) + (fmap->nareas * sizeof (FuFmap));
+}
+
+/* brute force linear search */
+static gboolean
+fmap_lsearch (const guint8 *image, gsize len, gsize *offset, GError **error)
+{
+	gsize i;
+	gboolean fmap_found = FALSE;
+
+	if (offset == NULL) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "offset return not valid");
+		return FALSE;
+	}
+
+	for (i = 0; i < len - strlen(FMAP_SIGNATURE); i++) {
+		if (!memcmp(&image[i],
+		            FMAP_SIGNATURE,
+		            strlen(FMAP_SIGNATURE))) {
+			fmap_found = TRUE;
+			break;
+		}
+	}
+
+	if (!fmap_found) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "fmap not found using linear search");
+		return FALSE;
+	}
+
+	if (i + fmap_size ((FuFmap *)&image[i]) > len) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "malformed fmap too close to end of image");
+
+		return FALSE;
+	}
+
+	*offset = i;
+	return TRUE;
+}
+
+/* if image length is a power of 2, use binary search */
+static gboolean
+fmap_bsearch (const guint8 *image, gsize len, gsize *offset, GError **error)
+{
+	gsize i;
+	gint stride;
+	gboolean fmap_found = FALSE;
+
+	if (offset == NULL) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "offset return not valid");
+		return FALSE;
+	}
+
+	/*
+	 * For efficient operation, we start with the largest stride possible
+	 * and then decrease the stride on each iteration. Also, check for a
+	 * remainder when modding the offset with the previous stride. This
+	 * makes it so that each offset is only checked once.
+	 */
+	for (stride = len / 2; stride >= 1; stride /= 2) {
```

Comment:
maybe move `gint stride` into the for loop.

Reply:
Done.